public/app/plugins/panel/geomap/layers/data/circlesOverlay.ts

```diff
@@ -0,0 +1,185 @@
+import { MapLayerRegistryItem, MapLayerConfig, MapLayerHandler, PanelData, GrafanaTheme2 } from '@grafana/data';
+import { dataFrameToLocations } from './utils'
+import { FieldMappingOptions, QueryFormat } from '../../types'
+import Map from 'ol/Map';
+import Feature from 'ol/Feature';
+import * as layer from 'ol/layer';
+import * as source from 'ol/source';
+import * as style from 'ol/style';
+import {Point} from 'ol/geom';
+import { fromLonLat } from 'ol/proj';
+import tinycolor from 'tinycolor2';
+
+export interface CircleConfig {
+  queryFormat: QueryFormat,
+  fieldMapping: FieldMappingOptions,
+  minSize: number,
+  maxSize: number,
+  opacity: number,
+}
+
+const defaultOptions: CircleConfig = {
+  queryFormat: {
+    locationType: 'coordinates',
+  },
+  fieldMapping: {
+    metricField: '',
+    geohashField: '',
+    latitudeField: '',
+    longitudeField: '',
+  },
+  minSize: 1,
+  maxSize: 10,
+  opacity: 0.4,
+};
+
+export const circlesLayer: MapLayerRegistryItem<CircleConfig> = {
+  id: 'circles',
+  name: 'Circles',
+  description: 'creates circle overlays for data values',
+  isBaseMap: false,
+
+  /**
+   * Function that configures transformation and returns a transformer
+   * @param options
+   */
+  create: (map: Map, options: MapLayerConfig<CircleConfig>, theme: GrafanaTheme2): MapLayerHandler => {
+    const config = { ...defaultOptions, ...options.config };
+
+    const vectorLayer = new layer.Vector({});
+    return {
+      init: () => vectorLayer,
+      update: (data: PanelData) => {
+        const features: Feature[] = [];
+
+        // Get data values
+        const dataArray = dataFrameToLocations(data.series[0], config);
+
+        // Map each data value into new points
+        dataArray.dataValues.map((datapoint) => {
+          // Get circle colors from threshold
+          const color = getColor(data.series[0].fields[0].config, theme, datapoint.value);
```

Comment:
This would be much nicer if we used the existing fields structure to get color -- it would also greatly simplify some of the gnarly bits here.

To do that we will need to revisit `dataFrameToLocations`

Reply:
"gnarly bits" is not fair -- it is just the `getColor` function :)